Bootstrapless addon

meandavejustice opened the pull request:
Rewrite of addon as webextension

Functionality:
- add button to toolbar which links to testpilot based on the currentEnv (local/prod etc...)
- toolbar link passes correct utm params
- show notification when new experiments or Major news updates are added to the site.
- submit a daily ping with a list of currently installed experiments

Structure:
- got rid of all preprocessing of files
- everything is in addon/background.js or options/options.js
- npm scripts in package.json


Issues:
- fixes #3382
- fixes #3183
- fixes #3176
- fixes #3177
- fixes #3182
- fixes #3180
- fixes #3313

Commit:
Remove appStartup

- refs #3182

Commit:
Remove Frame scripts

- fixes #3176

Commit:
Add options page for environment handling

- options page is not showing up yet, I have a hunch that this may
be because of the addon is still a bootstrap hyrid, something seems
to be blocking the loading of the options.html page. Going to keep
investigating, but wanted to commit to get these changes pushed to
remote.

Commit:
Copy over options to build dir in webpack.config

Commit:
Full conversion of addon to webextension

- fixes #3382
- fixes #3183
- fixes #3176
- fixes #3177
- fixes #3182
- fixes #3180
- fixes #3313

meandavejustice commented on addon/background.js:
```diff
@@ -0,0 +1,206 @@
+  });
+}
+
+async function setup() {
```
This is the starting point for the webextension.

Here we:
* fetch or generate a clientUUID(used to daily metric ping)
* setupEnvironment(setup listener for changing of addon environement(local/dev/stage/prod)
* setup our browser action button
* populate resources (experiments and news feed)
* detect our installed test pilot addons and add listeners to keep that updated on install/uninstall events.
* setup our daily ping interval

meandavejustice commented on addon/background.js:
```diff
@@ -0,0 +1,206 @@
+  });
+}
+
+function submitPing(object, event, addons) {
```
Telemetry and ping centre are no longer called in the new webextension. We're only using google analytics now.

meandavejustice commented on addon/background.js:
```diff
@@ -0,0 +1,206 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
This is where everything happens for the webextension.  Most of what is in there was pulled from `src/webextension/lib/{browserAction, storage}` and `src/lib/pings`.

meandavejustice commented on addon/bin/update-version:
```diff
@@ -1,43 +1,36 @@
-var manifestJSON = JSON.stringify(manifest, null, '  ');
-fs.writeFileSync(manifestFilename, manifestJSON);
-
-var installFilename = __dirname + '/../build/install.rdf';
```
This file is virtually the same except for generation of `install.rdf` was removed.

meandavejustice commented on addon/manifest.json:
```diff
@@ -16,6 +17,7 @@
   },
   "default_locale": "en_US",
   "permissions": [
+    "management",
```
used for tracking addon installs/uninstalls

dannycoates replied:
needs "alarms" permission

meandavejustice commented on addon/options/options.js:
```diff
@@ -0,0 +1,26 @@
+/* global browser */
```
Environments came from `src/webextension/lib/environments.js`. They are now part of the options page as we no longer have access to `about:config` prefs.

meandavejustice commented on addon/src/chrome/scripts/frame-script.js:
```diff
@@ -1,108 +0,0 @@
-/*
```
No longer use frame scripts for detecting test pilot addon on certain mozilla properties

meandavejustice commented on addon/src/install.rdf:
```diff
@@ -1,26 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
```
bootstrap-removal

meandavejustice commented on addon/src/lib/addonManager.js:
```diff
@@ -1,66 +0,0 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
```
now using the `management` api to fetch addons and determine which txp addons are installed at the time of the daily ping.

meandavejustice commented on addon/src/lib/appStartup.js:
```diff
@@ -1,30 +0,0 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
```
bootstrap-removal, `sessionstore-windows-restored` was a hack to speed up startup for hybrid bootstrap embedded webextensions.

meandavejustice commented on addon/src/lib/channels.js:
```diff
@@ -1,176 +0,0 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
```
bootstrap-removal, no longer provide metrics channel for experiments

meandavejustice commented on addon/src/lib/events.js:
```diff
@@ -1,87 +0,0 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
```
bootstrap-removal, this was a bridge for talking to the embedded webextension

meandavejustice commented on addon/src/lib/frameScripts.js:
```diff
@@ -1,63 +0,0 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
```
removal of frame scripts, see https://github.com/mozilla/testpilot/pull/3456/files#r178343655

meandavejustice commented on addon/src/lib/pings.js:
```diff
@@ -1,100 +0,0 @@
-// Shared metrics ping formatting & logic for bootstrap and webextension, with
```
ga ping snippet pulled for `background.js`. Everything else removed.

meandavejustice commented on addon/src/lib/prefs.js:
```diff
@@ -1,53 +0,0 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
```
Bootstrap-removal, no longer using prefs

meandavejustice commented on addon/src/lib/telemetry.js:
```diff
@@ -1,126 +0,0 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
```
bootstrap-removal, no more telemetry

meandavejustice commented on addon/src/lib/topics.js:
```diff
@@ -1,65 +0,0 @@
-// Shared list of topics for PubSub usage in bootstrap and webextension
```
bootstrap-removal, part of pubsub bridge to webextension. no longer needed

meandavejustice commented on addon/src/lib/utils.js:
```diff
@@ -1,17 +0,0 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
```
log is commented out for development use in `background.js`

meandavejustice commented on addon/src/lib/webExtension.js:
```diff
@@ -1,83 +0,0 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
```
bootstrap-removal, embedded webextension management and bridge

meandavejustice commented on addon/src/webextension/lib/browserAction.js:
```diff
@@ -1,102 +0,0 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
```
almost this whole file is now in `background.js`

meandavejustice commented on addon/src/webextension/background.js:
```diff
@@ -1,36 +0,0 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
```
bootstrap-removal

meandavejustice commented on addon/src/webextension/lib/bootstrap.js:
```diff
@@ -1,28 +0,0 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
```
bootstrap-removal

meandavejustice commented on addon/src/webextension/lib/environments.js:
```diff
@@ -1,98 +0,0 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
```
now in `options/options.js`

meandavejustice commented on addon/src/webextension/lib/metrics.js:
```diff
@@ -1,148 +0,0 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
```
bootstrap-removal, daily-ping was written fresh using browser.alarms in background.js

meandavejustice commented on addon/src/webextension/lib/storage.js:
```diff
@@ -1,23 +0,0 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
```
now in background.js

meandavejustice commented on addon/src/webextension/lib/utils.js:
```diff
@@ -1,14 +0,0 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
```
bootstrap-removal

meandavejustice commented on addon/test/test-setup.js:
```diff
@@ -1 +0,0 @@
-// Global setup for all tests
```
empty file

meandavejustice commented on addon/webpack.config.js:
```diff
@@ -1,109 +0,0 @@
-      // HACK: Only perform bundling if the second half of build has been
-      // performed, but we still want to rebuild & post the first half on file
-      // watching.
-      const buildPackagePath = './build/package.json';
```
moved into npm scripts

Commit:
Remove update-version from prepackage script

Commit:
Attempting fix for integration test, detecting web ext notifications on install

dannycoates commented on addon/background.js:
```diff
@@ -0,0 +1,206 @@
+  browser.management.onEnabled.addListener(setInstalledTxpAddons);
+  browser.management.onDisabled.addListener(setInstalledTxpAddons);
+  browser.management.onInstalled.addListener(setInstalledTxpAddons);
+  browser.management.onUninstalled.addListener(setInstalledTxpAddons);
```
It looks like we only use `installedTxpAddons` in the daily ping. It might be more polite to just run `setInstalledTxpAddons` then. Not a blocker. Your call.

dannycoates commented on addon/background.js:
```diff
@@ -0,0 +1,206 @@
+  browser.storage.onChanged.addListener((changes) => {
+    Object.keys(changes).forEach((k) => {
+      if (k === "environment") {
+        currentEnvironment = changes[k];
```
should be `changes[k].newValue`

dannycoates commented on addon/background.js:
```diff
@@ -0,0 +1,206 @@
+  const delayInMinutes = 5;
+  const periodInMinutes = 60 * 24; // daily
+
+  browser.alarms.create("daily-ping", {
```
So this will fire after 5 minutes every time I restart the browser, which might be ok? (not my call) If we really only want at most one ping per day we should save a timestamp to storage so we can check it on startup.

dannycoates commented on addon/background.js:
```diff
@@ -0,0 +1,206 @@
+  log("fetchResources");
+  return Promise.all(
+    Object.keys(resources).map(path =>
+                               fetch(`${currentEnvironment.baseUrl}/api/${path}.json`)
```
nit: this indenting is hard to read

dannycoates commented on addon/background.js:
```diff
@@ -0,0 +1,206 @@
+    text: (newExperiments.length || newsUpdates.length) > 0 ? "!" : ""
+  });
+
+  browser.browserAction.setBadgeText({
```
idk why this is here twice. bad merge?

dannycoates commented on addon/options/options.html:
```diff
@@ -0,0 +1,21 @@
+        <option value="https://testpilot.dev.mozaws.net">Development</option>
+        <option value="https://testpilot-l10n.dev.mozaws.net">L10n</option>
+        <option value="https://testpilot.stage.mozaws.net">Stage</option>
+        <option value="https://testpilot.firefox.com">Production</option>
```
since production is the default in background.js I think it should be the one initialized to selected here

Commit:
next pass as fixing install integration test

dannycoates commented on addon/background.js:
```diff
@@ -0,0 +1,206 @@
+  const twoWeeksAgo = Date.now() - TWO_WEEKS;
+  const newsUpdates = (news_updates || []).filter((u) => u.major)
+        .filter((u) => new Date(u.published).getTime() >= twoWeeksAgo)
+        .filter((u) => new Date(u.published).getTime() >= new Date(lastViewed))
```
nit: `lastViewed` shouldn't need a Date wrapper

Commit:
try same notification as experiment install

dannycoates commented on addon/background.js:
```diff
@@ -0,0 +1,206 @@
+async function setup() {
+  const data = await storage.get("clientUUID");
+  if (!data.clientUUID) {
+    await storage.set({ clientUUID: clientUUID = uuidv4() });
```
your `let clientUUID` above doesn't get initialized once it's already stored

dannycoates commented on addon/background.js:
```diff
@@ -0,0 +1,206 @@
+}
+
+const storage = browser.storage.local;
+const RESOURCE_UPDATE_INTERVAL = 10000; // 4 hours
```
`10000` isn't actually 4 hours ;)

meandavejustice replied:
oh that was a testing mistake. forgot to change, thanks for catching that!

Commit:
Address review comments

- remove unneccessary date wrapper
- fix daily ping to test date before sending request
- indent cleanup
- change default in options.html to local
- fix setting of environment from onchanged listener

Commit:
add alarm permission, clientUUID storage, resource fetch interval

dannycoates commented on addon/background.js:
```diff
@@ -0,0 +1,207 @@
+                                              (c ^ crypto.getRandomValues(new Uint8Array(1))[0] & 15 >> c / 4).toString(16));
+}
+
+function setInstalledTxpAddons() {
```
nit: it might be better to call this `getInstalledTxpAddons` now

dannycoates commented on addon/background.js:
```diff
@@ -0,0 +1,207 @@
+      const ONE_DAY = 60 * 60 * 1000 * 24; /* ms */
+      const { lastPing } = await storage.get("lastPing");
+      if (initial || ((new Date) - lastPing) > ONE_DAY) {
+        setInstalledTxpAddons().then((installedTxpAddons) => {
```
nit: `await` instead of `then`

dannycoates commented on addon/background.js:
```diff
@@ -0,0 +1,207 @@
+    periodInMinutes
+  });
+
+  browser.alarms.onAlarm.addListener((alarmInfo) => {
```
needs `async` on the callback

dannycoates commented on addon/background.js:
```diff
@@ -0,0 +1,207 @@
+  setupEnvironment();
+  setupBrowserAction();
+  await fetchResources();
+  setDailyPing();
```
should have `await` for consistency

dannycoates commented on addon/background.js:
```diff
@@ -0,0 +1,207 @@
+  const { clientUUID } = await storage.get("clientUUID");
+  if (!clientUUID) {
+    await storage.set({ clientUUID: uuidv4() });
+  }
```
putting all the clientUUID code in `setDailyPing` is more cohesive

dannycoates commented on addon/background.js:
```diff
@@ -0,0 +1,207 @@
+      if (initial || ((new Date) - lastPing) > ONE_DAY) {
+        setInstalledTxpAddons().then((installedTxpAddons) => {
+          const { clientUUID } = await storage.get({ clientUUID: uuidv4() });
+          submitPing(alarmInfo.name, "installed-addons", installedTxpAddons, clientUUID);
```
`lastPing` needs to be set after one is sent

dannycoates commented on addon/background.js:
```diff
@@ -0,0 +1,207 @@
+  if (!lastPing) {
+    initial = true;
+    await storage.set({ lastPing: Date.now()});
+  }
```
we can remove `initial` and skip this if we default `lastPing` to yesterday

meandavejustice replied:
woah clever :+1:

Commit:
organize dailyPing, clientUUID, set=>getInstalltxpAddons

Commit:
fix indentation for flake8

dannycoates commented on addon/background.js:
```diff
@@ -52,44 +53,46 @@ function setInstalledTxpAddons() {
+
+async function sendPingOnAlarm(alarmInfo) {
+  if (alarmInfo.name === "daily-ping") {
+    const { lastPing } = await storage.get("lastPing");
```
I was thinking more along the lines of `storage.get({ lastPing: Date.now() - ONE_DAY })`

meandavejustice replied:
I don't think this would ever return successfully :thinking:

Commit:
add confirmation notifcation handling to experiment page test

dannycoates commented on addon/background.js:
```diff
@@ -52,44 +53,46 @@ function setInstalledTxpAddons() {
-      }
-    }
-  });
+  browser.alarms.onAlarm.addListener(sendPingOnAlarm);
```
nice

Commit:
cleanup lastPing defaults and remove async from setDailyPing
